### ISSUE TITLE
Proper way to copy ofTexture to ofImage?

### ISSUE THREAD

**Issue by cuinjune** [openframeworks/openFrameworks#5669]
Hi, I'm trying to copy font's texture to ofImage.
I found this post about how to copy ofTexture to ofImage
https://forum.openframeworks.cc/t/oftexture-to-ofimage/4451/6

So I tried this based on @polyclick's suggestion

    void ofApp::setup(){

        font.load("font.ttf", 50);
        ofPixels pixels;
        font.getFontTexture().readToPixels(pixels);
        image.setFromPixels(pixels); 
    }
But when I run it, I see these errors on the console.
> [ error ] ofPixels: image type not supported
> [ error ] ofPixels: format doesn't support channels
> [ error ] ofGLUtils: ofGetGLFormatFromPixelFormat(): unknown OF pixel format-1, returning GL_LUMINANCE
<img width="1139" alt="a8b1bf20fbf481ef9f7f4f74e486c8f9a9ea3f12" src="https://user-images.githubusercontent.com/24947986/27160029-10aa8466-51ac-11e7-9a42-da15e1d3ba8b.png">

How can I properly copy font Texture to ofImage?
I use OF 0.9.8 on OSX.
Thank you so much in advance!

**Comment by bakercp** [comment]
I can confirm that this occurs in the master branch on OSX 10.12.5.

There is a bug in here somewhere and a lot of seems to come down to a limitation of `ofImage` / `ofImageType`.

The underlying font's texture atlas is a 2 channel grayscale alpha image (`OF_PIXELS_GRAY_ALPHA`), which has no corresponding `ofImageType` , thus it says it's unsupported.  Any conversion function such as `ofImageType ofImageTypeFromPixelFormat(ofPixelFormat pixelFormat)` or `ofPixelFormat ofPixelFormatFromImageType(ofImageType type)` are losing information about the underlying data types.

Thus it is certainly possible to do a readback to ofPixels if done manually.  Here is a re-write of the `getFontTexture().readToPixels(...)` function that correctly translates the texture type to an `ofPixelType` (rather than the ofImageType). 

```c++
    const ofTexture& tex = font.getFontTexture();

    auto& texData = tex.texData;

    pixels.allocate(texData.width,
                    texData.height,
                    OF_PIXELS_GRAY_ALPHA); /// <-- manually set this.

    ofSetPixelStoreiAlignment(GL_PACK_ALIGNMENT,
                              pixels.getWidth(),
                              pixels.getBytesPerChannel(),
                              pixels.getNumChannels());

    glBindTexture(texData.textureTarget,texData.textureID);

    glGetTexImage(texData.textureTarget,
                  0,
                  ofGetGlFormat(pixels),
                  GL_UNSIGNED_BYTE,
                  pixels.getData());

    glBindTexture(texData.textureTarget, 0);
```

Bottom line is, we need to prevent the crash for sure and perhaps we need to re-align ofImage / and ofImageType with ofPixels since ofPixels has been significantly improved and updated to work better with the programmable renderer while ofImage has lagged behind a bit.

**Comment by bakercp** [comment]
Here are some past issues that further this discussion:

https://github.com/openframeworks/openFrameworks/pull/2759
https://github.com/openframeworks/openFrameworks/issues/2854
https://github.com/openframeworks/openFrameworks/issues/2502

@arturoc
@tgfrerer

**Comment by mimetikxs** [comment]
Hi guys, I'm also having problems with this method.

This code:
```
// setup
fbo.allocate(w, h, GL_R8);
pixels.allocate(w, h, OF_PIXELS_R);

// update
fbo.readToPixels(pixels);
```
Outputs this error:
> [ error ] ofPixels: image type not supported

But the application works normally. 

The problem seems to be that`ofGetImageTypeFromGLType` is not checking `GL_R8` in the switch statement:

```
ofImageType ofGetImageTypeFromGLType(int glType){
	switch(glType){
	case GL_LUMINANCE:
#ifndef TARGET_OPENGLES
	case GL_LUMINANCE8:
	case GL_LUMINANCE16:
	case GL_LUMINANCE32F_ARB:
	case GL_R16:
	case GL_R16F:
	case GL_R16I:
	case GL_R16UI:
	case GL_R32F:
	case GL_R32I:
	case GL_R32UI:
	case GL_DEPTH_COMPONENT32F:
	case GL_DEPTH_COMPONENT32:
	case GL_DEPTH_COMPONENT16:
	case GL_DEPTH_COMPONENT24:
	case GL_DEPTH_COMPONENT:
#endif
		return OF_IMAGE_GRAYSCALE;
```
 @arturoc mentioned in [this comment](https://github.com/openframeworks/openFrameworks/issues/2854#issuecomment-37724116) that he added the 8bit types but GL_R8 is still missing.
